### PR TITLE
Move tools to a plugin library

### LINKED ISSUE #301: Simple dependency management between plugins
## The Problem
We want to create a new plugin, that combines data from WMS, WCS and WFS sources that also implements some analysis tools ontop. We currently have a plugin to load WMS data and [a WCS plugin is also in the vestec branch](https://github.com/cosmoscout/cosmoscout-vr/tree/feature/vestec/plugins/csp-wcs-overlays). 

Currently it is not easily possible to have dependencies between plugins, which initially was also by design, but it would also be code smell to duplicate all this plugin code.

## The Solution
Create a small dependency system, that would allow plugins to depend on other plugins.

Things to consider:
- Plugin dependencies have to be declared in the config
- A dependency tree has to be created and the plugins have to be loaded in the correct order
- Plugins should have a headless mode, where the DLL is loaded, but the plugin is not initialized, so they function more like a library

## The Alternatives
- Copy and paste the required code to the new plugin
- Put some of the wms/wcs/wfs features into CosmoScout core
- Create some kind of data sharing feature in CosmoScout core

### PR DESCRIPTION
**Pull request by JonasGilg**
This PR adds a new type of plugin called a plugin library. Plugin libraries are also located in the plugins folder, but are prefixed with `csl`, instead of `csp`.

- With the introduction of this feature the tools were moved from `cs-core` to `csl-tools`.
- To easily create plugin libraries a new PowerShell script was added: `tools/New-PluginLibrary.ps1`

I tested the loading of `csl-tools` with a copy of measurement tools on Windows. I did the following tests:
- Load both plugins and make sure `csl-tools` is loaded only once.
- Unload one plugin and make sure `csl-tools` is not unloaded for the other plugin.
- Unload both plugins and make sure `csl-tools` is also unloaded.
- Load both plugins again. A new instance of `csl-tools` was loaded.

To ensure that `csl-tools` was only loaded once, I logged the memory address of the `logger`.

Resolves #301

### CODE REVIEW

File: plugins/csl-tools/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+file(GLOB SOURCE_FILES src/*.cpp)
+
+# Resource files and header files are only added in order to make them available in your IDE.
+file(GLOB HEADER_FILES src/*.hpp)
+
+add_library(csl-tools SHARED
+  ${SOURCE_FILES}
+  ${HEADER_FILES}
+  ${RESOURCE_FILES}
+)
```
Comment (Schneegans) [inline plugins/csl-tools/CMakeLists.txt:25]: There are no resource files here.
```suggestion
  ${HEADER_FILES}
)
```

File: plugins/csl-tools/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# Add this Plugin to a "plugins" folder in your IDE.
+set_property(TARGET csl-tools PROPERTY FOLDER "plugins")
+
+# We mark all resource files as "header" in order to make sure that no one tries to compile them.
+set_source_files_properties(${RESOURCE_FILES} PROPERTIES HEADER_FILE_ONLY true)
+
+# Make directory structure available in your IDE.
+source_group(TREE "${CMAKE_CURRENT_SOURCE_DIR}" FILES
+  ${SOURCE_FILES} ${HEADER_FILES} ${RESOURCE_FILES}
+)
```
Comment (Schneegans) [inline plugins/csl-tools/CMakeLists.txt:42]:
```suggestion
# Make directory structure available in your IDE.
source_group(TREE "${CMAKE_CURRENT_SOURCE_DIR}" FILES
  ${SOURCE_FILES} ${HEADER_FILES}
)
```

File: plugins/csl-tools/src/Mark.cpp [hunk 3]
```diff
@@ -7,15 +7,17 @@
...
+#include "../../../src/cs-core/InputManager.hpp"
+#include "../../../src/cs-core/Settings.hpp"
+#include "../../../src/cs-core/SolarSystem.hpp"
+#include "../../../src/cs-core/TimeControl.hpp"
+#include "../../../src/cs-scene/CelestialObject.hpp"
+#include "../../../src/cs-scene/CelestialSurface.hpp"
+#include "../../../src/cs-utils/convert.hpp"
+#include "../../../src/cs-utils/utils.hpp"
+
+#include "logger.hpp"
```
Comment (Schneegans) [inline plugins/csl-tools/src/Mark.cpp:20]: I think this can be removed again.
```suggestion
```

File: plugins/csl-tools/src/Mark.cpp [hunk 4]
```diff
@@ -190,6 +193,8 @@ bool Mark::GetBoundingBox(VistaBoundingBox& bb) {
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
 void Mark::initData() {
+  logger().info("Initializing Mark");
+
```
Comment (Schneegans) [inline plugins/csl-tools/src/Mark.cpp:197]:
```suggestion
```

File: plugins/csl-tools/src/logger.cpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#include "../../../src/cs-utils/logger.hpp"
+
+namespace csl::tools {
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+spdlog::logger& logger() {
+  static auto logger = cs::utils::createLogger("csl-tools");
+  logger->info("{}", reinterpret_cast<size_t>(logger.get()));
+  return *logger;
```
Comment (Schneegans) [inline plugins/csl-tools/src/logger.cpp:19]: This was nice for testing :wink: 
```suggestion
  static auto logger = cs::utils::createLogger("csl-tools");
  return *logger;
```

File: tools/New-PluginLibrary.ps1 [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+# SPDX-License-Identifier: MIT
+
+<#
+    .SYNOPSIS
+    This function creates a new CosmoScout VR plugin library with the given name.
+    
+    .DESCRIPTION
+    This function creates a new CosmoScout VR plugin library with the given name. It will generate
+    all relevant files for
+    your Plugin:
```
Comment (Schneegans) [inline tools/New-PluginLibrary.ps1:15]:
```suggestion
    your plugin library:
```

File: tools/New-PluginLibrary.ps1 [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    - README.md
+    - CMakeLists.txt
+    - src/logger.hpp
+    - src/logger.cpp
+    
+    .PARAMETER Name
+    The plugin name. Just write it out in natural language. Don't use CamelCase, KebabCase,
```
Comment (Schneegans) [inline tools/New-PluginLibrary.ps1:22]:
```suggestion
    The plugin library name. Just write it out in natural language. Don't use CamelCase, KebabCase, 
```

File: tools/New-PluginLibrary.ps1 [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+file(GLOB SOURCE_FILES src/*.cpp)
+
+# Resource files and header files are only added in order to make them available in your IDE.
+file(GLOB HEADER_FILES src/*.hpp)
+
+add_library(csl-$kebabCaseName SHARED
+  `${SOURCE_FILES}
+  `${HEADER_FILES}
+  `${RESOURCE_FILES}
```
Comment (Schneegans) [inline tools/New-PluginLibrary.ps1:98]: Should we remove the resource files here as well?

**Review by Schneegans — CHANGES_REQUESTED** [review]
Thank you very much! This looks very good! I did not test the script, but I tested all the loading and unloading of an additional `csp-measurement-tools` plugin as you described but on Linux. Everything works as expected!

File: tools/New-PluginLibrary.ps1 [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+# Add this plugin library to a `"plugins`" folder in your IDE.
+set_property(TARGET csl-$kebabCaseName PROPERTY FOLDER `"plugins`")
+
+# We mark all resource files as `"header`" in order to make sure that no one tries to compile them.
+set_source_files_properties(`${RESOURCE_FILES} PROPERTIES HEADER_FILE_ONLY true)
+
+# Make directory structure available in your IDE.
+source_group(TREE `"`${CMAKE_CURRENT_SOURCE_DIR}`" FILES
+  `${SOURCE_FILES} `${HEADER_FILES} `${RESOURCE_FILES}
+)
```
Comment (JonasGilg) [inline tools/New-PluginLibrary.ps1:115]:
```suggestion
# Make directory structure available in your IDE.
source_group(TREE `"`${CMAKE_CURRENT_SOURCE_DIR}`" FILES
  `${SOURCE_FILES} `${HEADER_FILES}
)
```

**Review by Schneegans — APPROVED** [review]

**Pull request merged**